File: backend/app/models/attendance.py
import uuid
from sqlalchemy import Column, String, DateTime, ForeignKey, Text, UniqueConstraint
from sqlalchemy.sql import func
from ..database import Base

def generate_uuid():
    return str(uuid.uuid4())

class Attendance(Base):
    __tablename__ = "attendances"

    id = Column(String, primary_key=True, index=True, default=generate_uuid)
    employee_id = Column(String, ForeignKey("users.id"), nullable=False, index=True)
    
    date = Column(String, nullable=False, index=True) # YYYY-MM-DD format
    status = Column(String, default="present") # present, absent, late, half_day, work_from_home, on_leave
    check_in = Column(String, default="") # HH:MM format
    check_out = Column(String, default="") # HH:MM format
    notes = Column(Text, default="")
    marked_by = Column(String, ForeignKey("users.id"), nullable=True) # Who marked it (id)

    __table_args__ = (
        UniqueConstraint('employee_id', 'date', name='uq_employee_date'),
    )

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

File: ims-platform/backend/src/models/Attendance.js
'use strict';

const mongoose = require('mongoose');

const attendanceSchema = new mongoose.Schema(
    {
        employeeId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
        date: { type: String, required: true, index: true }, // YYYY-MM-DD
        status: {
            type: String,
            enum: ['present', 'absent', 'late', 'half_day', 'work_from_home', 'on_leave'],
            default: 'present',
        },
        checkIn: { type: String, default: '' },   // HH:MM
        checkOut: { type: String, default: '' },  // HH:MM
        notes: { type: String, default: '' },
        markedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    },
    { timestamps: true }
);

// Unique per employee per day
attendanceSchema.index({ employeeId: 1, date: 1 }, { unique: true });

module.exports = mongoose.model('Attendance', attendanceSchema);
